### PR TITLE
2844 lesmis users table part 2

### PR DESCRIPTION
**Pull request by tcaiger**
### Changes:

- Add data grid components
- Add UsersView with data grid
- Add LesmisPrivateRoute component and check if user is lesmis admin
- Add new NotAuthorisedView and NotFoundView

---

### Screenshots:
See issue

### CODE REVIEW

File: packages/lesmis/src/routes/LesmisAdminRoute.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ *
+ */
+import React from 'react';
+import { Route } from 'react-router-dom';
+import { useUser } from '../api/queries';
+import { FullPageLoader } from '../components';
+import { NotAuthorisedView } from '../views/NotAuthorisedView';
+
+export const LesmisAdminRoute = props => {
```
Comment (edmofro) [inline packages/lesmis/src/routes/LesmisAdminRoute.js:12]: Love this component

File: packages/lesmis/src/components/DataGrid/DataGrid.js [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ *
+ */
+import React from 'react';
+import PropTypes from 'prop-types';
+import styled from 'styled-components';
+import { useTable, useFilters, useSortBy, usePagination } from 'react-table';
+import { Table, TableHead, TableRow, TableBody, TableSortLabel } from '@material-ui/core';
+import { TableCell, StyledTableRow } from '@tupaia/ui-components';
```
Comment (edmofro) [inline packages/lesmis/src/components/DataGrid/DataGrid.js:11]: What it's importing from where seems a little hard to work out if I were creating another data grid somewhere else. I need most of the Table stuff from mui, but the TableCell (unstyled) and StyledTableRow (styled) from ui-components. 

Personally don't think any of the styling used in LESMIS is very LESMIS-specific, and we could pull the DataGrid up to the ui-components layer?

File: packages/lesmis/src/components/DataGrid/DataGrid.js [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+const StyledTableCell = styled(TableCell)`
+  padding: 1.125rem 0.625rem 1.125rem 0;
+  background: ${props => props.theme.palette.grey['200']};
+`;
+
+const filterTypes = () => ({
+  text: (rows, id, filterValue) => {
+    return rows.filter(row => {
+      const rowValue = row.values[id];
```
Comment (edmofro) [inline packages/lesmis/src/components/DataGrid/DataGrid.js:29]:
```suggestion
      const cellValue = row.values[id];
```

File: packages/lesmis/src/components/DataGrid/DataGrid.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+  return (
+    <>
+      <StyledTable {...getTableProps()}>
+        <TableHead>
+          {headerGroups.map(headerGroup => (
+            <TableRow {...headerGroup.getHeaderGroupProps()}>
+              {headerGroup.headers.map(column => (
+                <TableCell {...column.getHeaderProps(column.getSortByToggleProps())}>
+                  {column.render('Header')}
```
Comment (edmofro) [inline packages/lesmis/src/components/DataGrid/DataGrid.js:79]: Wow this is an unusual api

File: packages/lesmis/src/components/DataGrid/DataGrid.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+              ))}
+            </TableRow>
+          ))}
+        </TableHead>
+        <TableBody {...getTableBodyProps()}>
+          {page.length === 0 ? (
+            <NoDataRow colspan={columns.length} />
+          ) : (
+            page.map(row => {
+              prepareRow(row);
```
Comment (edmofro) [inline packages/lesmis/src/components/DataGrid/DataGrid.js:103]: Haha it's definitely a bit funky this new `react-table` version

File: packages/lesmis/src/components/DataGrid/DataGrid.js [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+      firstName: PropTypes.string,
+      lastName: PropTypes.string,
+      email: PropTypes.string,
+      permissionGroupName: PropTypes.string,
+    }),
+  ),
+  columns: PropTypes.arrayOf(
+    PropTypes.shape({
+      accessor: PropTypes.string,
+      Header: PropTypes.string,
```
Comment (edmofro) [inline packages/lesmis/src/components/DataGrid/DataGrid.js:141]: Where does `canFilter` get applied? It doesn't appear in the shape of prop that can be passed through with each column, but that's where I'd expect it was controlled?

**Review by edmofro — CHANGES_REQUESTED** [review]
Awesome work getting that DataGrid all hooked up! Some weird stuff there but the end result is a really nice separation of concerns

Main note would be - can we move it to ui-components?

File: packages/lesmis/src/components/DataGrid/Paginator.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+  nextPage,
+  pageIndex,
+  pageSize,
+  previousPage,
+  setPageSize,
+  totalCount,
+}) => {
+  const pageStart = pageIndex * pageSize + 1;
+  const pageEnd = (pageIndex + 1) * pageSize;
+  const overallPageEnd = pageEnd > totalCount ? totalCount : pageEnd;
```
[the remark this answers is not in the document]
Comment (tcaiger) [reply packages/lesmis/src/components/DataGrid/Paginator.js:78]: Great suggestion - thanks

File: packages/lesmis/src/views/UsersView.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  return (
+    <>
+      <PageHeader
+        title="Users and Permissions"
+        breadcrumbs={[{ name: 'Users and Permissions', url: '/users-and-permissions' }]}
+      />
+      <Section>
+        <Container maxWidth="lg">
+          <FetchLoader isLoading={isLoading} isError={isError} error={error}>
+            <Text>{data?.length} users</Text>
```
[the remark this answers is not in the document]
Comment (tcaiger) [reply packages/lesmis/src/views/UsersView.js:53]: Yeah I thought a little random. I'd be keen to get Dave to do a quick design review of the lesmis admin section stuff once it's all done (probably next week).

File: packages/lesmis/src/components/DataGrid/DataGrid.js (hunk 6 above)
Comment (tcaiger) [reply packages/lesmis/src/components/DataGrid/DataGrid.js:141]: Yeah good question. The canFilter prop can be set in the columns or it is true if the column has a Filter component set which in our case is applied through the `defaultColumn`.

**Review by edmofro — APPROVED** [review]
Good to go!

**Pull request merged**